client/src/controllers/DrilldownController.test.js

```diff
@@ -0,0 +1,201 @@
+
+  const events = {};
+  let application;
+  let errors = [];
+
+  const setup = async (
+    html = `<section>
+              <div id="filters-drilldown" class="w-drilldown" data-controller="w-drilldown" data-action="w-swap:success@document->w-drilldown#updateCount w-dropdown:hide->w-drilldown#delayedClose w-dropdown:clickaway->w-drilldown#preventOutletClickaway" data-w-drilldown-count-attr-value="data-w-active-filter-id">
+                <span class="w-drilldown__count" data-w-drilldown-target="count"></span>
+                  <button type="button">Show Filters</button>
+                <div class="w-drilldown__menu" data-w-drilldown-target="menu">
+                  <h2>Filter by</h2>
+                  <button class="w-drilldown__toggle" type="button" aria-expanded="false" aria-controls="drilldown-field-0" data-w-drilldown-target="toggle" data-action="click->w-drilldown#open">Field 1</button>
+                </div>
+                <div class="w-drilldown__submenu" id="drilldown-field-0" hidden tabIndex="-1">
+                  <button class="w-drilldown__back" type="button" data-action="click->w-drilldown#close">Back</button>
+                </div>
+              </div>
+              <button id="action-element" data-controller="w-action">Action</button>
+            </section>`,
+    identifier = 'w-drilldown',
+  ) => {
+    document.body.innerHTML = `<main>${html}</main>`;
+
+    application = new Application();
+    application.handleError = (error, message) => {
+      errors.push({ error, message });
+    };
+    application.register(identifier, DrilldownController);
+    application.register('w-action', ActionController);
+    application.register('w-dropdown', DropdownController);
+    application.start();
+
+    await jest.runAllTimersAsync();
+
+    return [
+      ...document.querySelectorAll(`[data-controller~="${identifier}"]`),
+    ].map((element) =>
+      application.getControllerForElementAndIdentifier(element, identifier),
+    );
+  };
+
+  beforeEach(async () => {
+    await setup();
+  });
+
+  afterEach(() => {
+    application?.stop && application.stop();
+    errors = [];
+    eventNames.forEach((name) => {
+      events[name] = [];
+    });
+  });
+
+  describe('DrilldownController open/close actions', () => {
+    it('opens the submenu and sets active submenu value when toggle is clicked', async () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      const submenu = document.querySelector('#drilldown-field-0');
+
+      toggleButton.click();
+      await jest.runAllTimersAsync();
+
+      expect(toggleButton.getAttribute('aria-expanded')).toBe('true');
+      expect(submenu.hidden).toBeFalsy();
+    });
+
+    it('closes the submenu and clears active submenu value when close action is triggered', () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      toggleButton.click();
+
```

Comment:
When doing a test like this, think about structuring your tests with Arrange, Act, Assert.
https://docs.telerik.com/devtools/justmock/basic-usage/arrange-act-assert

In it's current form it's not clear **why** the test is running this first `click()`. I assume you want to 'open' the drilldown and test that closing it works.

In which case, this is really part of your 'arranging', you are getting the DOM into a place so you can check something else.

You can probably add a new assertion under this first click, checking that the thing is open (or ready), this way you know that the 'arrange' has been done.

Sometimes it helps to even have comments (but remove them before the commit).

```
    it('closes the submenu and clears active submenu value when close action is triggered', () => {
      // arrange
      const toggleButton = document.querySelector(
        '[data-w-drilldown-target="toggle"]',
      );
      toggleButton.click();
     // add an assertion here that confirms the DOM is correctly arranged, I assume
    expect(submenu.hidden).toBeFalsy();

      const closeButton = document.querySelector(
        '[data-action="click->w-drilldown#close"]',
      );

      // act
      closeButton.click();

      const submenu = document.querySelector('#drilldown-field-0');

      // assert
      expect(toggleButton.getAttribute('aria-expanded')).toBe('false');
      expect(submenu.hidden).toBeTruthy();
    });
```

Reply:
Here's another link with Jest examples.
https://robertmarshall.dev/blog/arrange-act-and-assert-pattern-the-three-as-of-unit-testing/